migration/workspace-query-separation/src/main/java/ddf/migration/workspace/query/separation/WorkspaceQueryMigration.java

```diff
@@ -0,0 +1,205 @@
+
+  public WorkspaceQueryMigration(
+      CatalogFramework catalogFramework,
+      FilterBuilder filterBuilder,
+      InputTransformer xmlInputTransformer) {
+    this.catalogFramework = catalogFramework;
+    this.filterBuilder = filterBuilder;
+    this.xmlInputTransformer = xmlInputTransformer;
+  }
+
+  @Override
+  public void migrate() {
+    Subject systemSubject = SECURITY.runAsAdmin(SECURITY::getSystemSubject);
+
+    if (systemSubject == null) {
+      LOGGER.debug("An error occurred while attempting to run this migration as admin");
+      return;
+    }
+
+    systemSubject.execute(this::migrateWorkspaces);
+  }
+
+  @VisibleForTesting
+  void migrateWorkspaces() {
+    LOGGER.debug("Beginning workspace query data migration");
+    System.out.println("Started: Workspace Query Data Migration");
+
+    Filter workspaceFilter =
+        filterBuilder.allOf(
+            filterBuilder.attribute(Core.METACARD_TAGS).like().text(WORKSPACE_TAG),
+            filterBuilder.attribute(WORKSPACE_QUERIES).is().like().text("*"));
+
+    Query query = new QueryImpl(workspaceFilter);
+    QueryRequest queryRequest = new QueryRequestImpl(query);
+
+    ResultIterable.resultIterable(catalogFramework::query, queryRequest)
+        .stream()
+        .map(Result::getMetacard)
+        .filter(Objects::nonNull)
+        .forEach(this::migrateWorkspaceMetacard);
+
+    LOGGER.debug("Completed workspace query data migration");
+    System.out.println("Completed: Workspace Query Data Migration");
+  }
+
+  private void migrateWorkspaceMetacard(Metacard workspaceMetacard) {
+    LOGGER.debug("Migrating workspace metacard with id [{}]", workspaceMetacard.getId());
+
+    final Attribute queriesAttribute = workspaceMetacard.getAttribute(WORKSPACE_QUERIES);
+    final List<Serializable> queriesAsXml = queriesAttribute.getValues();
+
+    LOGGER.debug(
+        "Beginning migration of {} query metacards from workspace [{}]",
+        queriesAsXml.size(),
+        workspaceMetacard.getId());
+
+    List<Metacard> queryMetacards =
+        queriesAsXml
+            .stream()
+            .filter(String.class::isInstance)
+            .map(String.class::cast)
+            .map(xml -> xml.getBytes(StandardCharsets.UTF_8))
+            .map(ByteArrayInputStream::new)
+            .map(this::xmlToMetacard)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+    LOGGER.debug("Creating query metacards for workspace [{}]", workspaceMetacard.getId());
+
+    CreateResponse response = createMetacards(queryMetacards);
+
+    if (response != null) {
+      List<String> createdMetacardIds =
+          response.getCreatedMetacards().stream().map(Metacard::getId).collect(Collectors.toList());
+
+      LOGGER.debug("Created query metacards with IDs {}", createdMetacardIds);
+
+      workspaceMetacard.setAttribute(
+          new AttributeImpl(WORKSPACE_QUERIES, (Serializable) createdMetacardIds));
```

Comment:
We can't safely cast the list returned by the collector to `Serializable` since there are no guarantees on the serializability of the `List` (see https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toList--).

You could change the type of `createdMetacardIds` to a `List<Serializable>` and remove the cast here. Or instead of `toList()` you could use `toCollection(ArrayList::new)` or `toCollection(LinkedList::new)` since those `List` implementations are serializable. I prefer the first option.